### DISCUSSION TITLE
Any way to compress/decompress multiple images at once?

### DISCUSSION THREAD

**Discussion by saltwick** [tensorflow/compression#128] [Q&A] [answered]
To compress/decompress on multiple images, I have made some change to load the model once then run the sender/receiver many times. Is there a way to pass a whole batch into the sender/receiver function?

**Comment by jonaballe** [comment, marked as answer]
Hi @saltwick,

yes, you can pass a batch of images into pre-trained models. A batch dimension is added here:
https://github.com/tensorflow/compression/blob/master/models/tfci.py#L45
and removed again here:
https://github.com/tensorflow/compression/blob/master/models/tfci.py#L50

If you pass a batch of images to the models in that dimension, they should return batched compressed strings. Note we haven't tested this approach, and this would naturally only work if the images are all the same size, which is one reason why we didn't implement this. Also, when using the `--target_bpp` parameter, you'd need to iterate separately over each image, as the R/D tradeoff will be different for each one.

If you can't make it work this way, please let me know.

**Reply by saltwick** [reply]
Would this work when decompressing as well? In the decompress function, the receiver function accepts `*tensors`. The `tensors` variable appears to be a list of 7 EagerTensor objects. To pass many into receiver, I'm not sure which of those 7 inputs need to change.

Edit: After looking more into the inputs I was able to accomplish this by concatenating on the first, second, sixth, and seventh inputs. Thanks for your help!

**Reply by jonaballe** [reply]
Yes, for the record, the tensors will be either shape tensors or string tensors containing compressed data. If you are batching, the shape tensors will be the same for all of the images, so they won't change, but the string tensors will be batched.

**Reply by gregg-p** [reply]
How would you go about passing a batch of images? What would the arguments look like in the terminal?

**Reply by jonaballe** [reply]
You would have to modify the command line parsing in the script to account for processing multiple images. The models themselves can process batches, but the tfci.py script is not designed to do that (mainly because most images are going to have differing sizes).

If you want to process multiple images right now, I'd suggest to just call tfci.py multiple times, e.g. using a for loop in the shell.

One convenient addition to the script could be to make it process multiple images, but not attempt to process them in batches. That way, loading the model itself multiple times could be prevented, and it might overall run somewhat faster.
